crates/oxc_regular_expression/src/display.rs

```diff
@@ -0,0 +1,586 @@
+        write!(f, "{}-{}", self.min, self.max)
+    }
+}
+
+impl<'a> Display for ClassStringDisjunction<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, r"\q{{")?;
+        write_join(f, "|", &self.body)?;
+        write!(f, "}}")
+    }
+}
+
+impl<'a> Display for ClassString<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_join(f, "", &self.body)
+    }
+}
+
+impl<'a> Display for CapturingGroup<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let body = &self.body;
+        if let Some(name) = &self.name {
+            write!(f, "(?<{name}>{body})")
+        } else {
+            write!(f, "({body})")
+        }
+    }
+}
+
+impl<'a> Display for IgnoreGroup<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fn write_flags(
+            f: &mut fmt::Formatter<'_>,
+            prefix: char,
+            flags: &ModifierFlags,
+        ) -> fmt::Result {
+            if flags.ignore_case {
+                write!(f, "{prefix}i")?;
+            }
+            if flags.sticky {
+                write!(f, "{prefix}y")?;
+            }
+            if flags.multiline {
+                write!(f, "{prefix}m")?;
+            }
+            Ok(())
+        }
+
+        write!(f, "(?")?;
+        if let Some(enabling) = &self.enabling_modifiers {
+            write_flags(f, '\0', enabling)?;
+        }
+        if let Some(disabling) = &self.disabling_modifiers {
+            write_flags(f, '-', disabling)?;
+        }
+        write!(f, ":{})", self.body)
+    }
+}
+
+impl Display for IndexedReference {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, r"\{}", self.index)
+    }
+}
+
+impl<'a> Display for NamedReference<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, r"\k<{}>", self.name)
+    }
+}
+
+fn write_join<S, I, E>(f: &mut fmt::Formatter<'_>, sep: S, items: I) -> fmt::Result
+where
+    S: AsRef<str>,
+    E: Display,
+    I: IntoIterator<Item = E>,
+{
+    write_join_with(f, sep, items, |iter| iter.next().map(|it| it.to_string()))
+}
+
+fn write_join_with<S, I, E, F>(f: &mut fmt::Formatter<'_>, sep: S, items: I, next: F) -> fmt::Result
+where
+    S: AsRef<str>,
+    E: Display,
+    I: IntoIterator<Item = E>,
+    F: Fn(&mut Peekable<I::IntoIter>) -> Option<String>,
+{
+    let sep = sep.as_ref();
+    let iter = &mut items.into_iter().peekable();
+
+    if let Some(first) = next(iter) {
+        write!(f, "{first}")?;
+    }
+
+    while let Some(it) = next(iter) {
+        write!(f, "{sep}{it}")?;
+    }
+
+    Ok(())
+}
+
+fn character_to_string(
+    this: &Character,
+    peek: Option<&Character>,
+) -> (/* result */ String, /* true of peek is consumed */ bool) {
+    const USED_PEEK: bool = true;
+    const UNUSED_PEEK: bool = false;
+
+    if is_trail_surrogate(this.value) {
```

Comment:
From my understanding, surrogate pair is only concern for `kind::Symbol`, so how about like this?

```rs
fn character_to_string(
    this: &Character,
    peek: Option<&Character>,
) -> (/* result */ String, /* true of peek should be consumed */ bool) {
    let cp = this.value;

    if let CharacterKind::Symbol = this.kind {
        // Trail only
        if is_trail_surrogate(cp) {
            return (format!(r"\u{cp:X}"), false);
        }

        if is_lead_surrogate(cp) {
            if let Some(peek) = peek.filter(|peek| is_trail_surrogate(peek.value)) {
                // Lead+Trail
                let cp = combine_surrogate_pair(cp, peek.value);
                let ch = char::from_u32(cp).expect("Invalid surrogate pair `Character`!");
                return (format!("{ch}"), true);
            }

            // Lead only
            return (format!(r"\u{cp:X}"), false);
        }
    }

    let ch = char::from_u32(cp).expect("Invalid `Character`!");
    let result = match this.kind {
        CharacterKind::ControlLetter => match ch {
            '\n' => r"\cJ".to_string(),
            '\r' => r"\cM".to_string(),
            '\t' => r"\cI".to_string(),
            _ => format!(r"\c{ch}"),
        },
        CharacterKind::Identifier => {
            format!(r"\{ch}")
        }
        // Not a surrogate, like BMP, or all units in unicode mode
        CharacterKind::Symbol => format!("{ch}"),
        CharacterKind::Null => String::from(r"\0"),
        CharacterKind::UnicodeEscape => {
            // we remove the leading `0x` of our 4 digit hex number.
            let hex = &format!("{cp:#4X}")[2..];
            if hex.len() <= 4 {
                format!(r"\u{hex}")
            } else {
                format!(r"\u{{{hex}}}")
            }
        }
        CharacterKind::HexadecimalEscape => {
            // we remove the leading `0x` of our 2 digit hex number.
            let hex = &format!("{cp:#2X}")[2..];
            format!(r"\x{hex}")
        }
        CharacterKind::Octal => {
            let octal = format!("{cp:o}");
            format!(r"\{octal}")
        }
        CharacterKind::SingleEscape => match ch {
            '\n' => String::from(r"\n"),
            '\r' => String::from(r"\r"),
            '\t' => String::from(r"\t"),
            '\u{b}' => String::from(r"\v"),
            '\u{c}' => String::from(r"\f"),
            '\u{8}' => String::from(r"\b"),
            '\u{2D}' => String::from(r"\-"),
            _ => format!(r"\{ch}"),
        },
    };

    (result, false)
}
```